### PR TITLE
video/fb: userspace framebuffer support

### PR DESCRIPTION
**Pull request by yf13**
## Summary

This enables user-space framebuffer access in kernel mode. It supports existing `examples/fb` on `rv-virt` device like below:

```
ABC
NuttShell (NSH) NuttX-12.4.0-RC0
nsh> cat /proc/version
NuttX version 12.4.0 1ea10ddacc-dirty May  3 2024 07:03:59 rv-virt/kfb64
nsh> fb
VideoInfo:
      fmt: 13
     xres: 640
     yres: 480
  nplanes: 1
PlaneInfo (plane 0):
    fbmem: 0x80218010
    fblen: 1228800
   stride: 2560
  display: 0
      bpp: 32
Mapped FB: 0xc2000010
 0: (  0,  0) (640,480)
 1: ( 58, 43) (524,394)
 2: (116, 86) (408,308)
 3: (174,129) (292,222)
 4: (232,172) (176,136)
 5: (290,215) ( 60, 50)
Test finished
nsh>
```

## Impact
userspace device mapping and video/fb driver

## Testing
with `rv-virt/kfb64` and CI checks

### CODE REVIEW

File: drivers/video/fb.c [hunk 1]
```diff
@@ -1032,7 +1032,13 @@ static int fb_mmap(FAR struct file *filep, FAR struct mm_map_entry_s *map)
   if (map->offset >= 0 && map->offset < panelinfo.fblen &&
       map->length && map->offset + map->length <= panelinfo.fblen)
     {
+#ifdef CONFIG_BUILD_KERNEL
+      map->vaddr = vm_map_region((uintptr_t)panelinfo.fbmem,
+                                            panelinfo.fblen);
+      map->vaddr += map->offset;
```
Comment (xiaoxiang781216) [inline drivers/video/fb.c:1038]: can vm_map_region support unaligned adress

File: drivers/virtio/virtio-gpu.c [hunk 2]
```diff
@@ -528,7 +535,8 @@ static int virtio_gpu_probe(FAR struct virtio_device *vdev)
   priv->stride = priv->xres * VIRTIO_GPU_BPP >> 3;
   priv->fblen  = priv->stride * priv->yres;
 
-  priv->fbmem  = (FAR uint8_t *)virtio_zalloc_buf(vdev, priv->fblen, 16);
+  priv->fbmem = (FAR uint8_t *)virtio_zalloc_buf(vdev, priv->fblen,
```
Comment (xiaoxiang781216) [inline drivers/virtio/virtio-gpu.c:538]: don't need change if vm_map_region support the unaligned address.

File: mm/map/Make.defs [hunk 3]
```diff
@@ -23,7 +23,7 @@
 CSRCS += mm_map.c
 
 ifeq ($(CONFIG_ARCH_VMA_MAPPING),y)
-CSRCS += vm_region.c
+CSRCS += vm_region.c vm_map.c
```
Comment (xiaoxiang781216) [inline mm/map/Make.defs:26]: why not put vm_map_region into vm_region.c?

**Review by acassis — APPROVED** [review]

File: drivers/video/fb.c (hunk 1 above)
Comment (yf13) [reply drivers/video/fb.c:1038]: I don't know how to support unaligned address yet. It seems that at RISC-V MMU requires aligned addresses, also granule allocator operates on aligned address as well.

File: mm/map/Make.defs (hunk 3 above)
Comment (yf13) [reply mm/map/Make.defs:26]: To make this patch focus on drivers layer, can we do that separately?

Comment (xiaoxiang781216) [reply mm/map/Make.defs:26]: OK

File: drivers/video/fb.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/video/fb.c:1038]: it's simple to move the offset addition to vm_map_region

File: mm/map/Make.defs (hunk 3 above)
Comment (yf13) [reply mm/map/Make.defs:26]: thanks, will do it later separately.

File: drivers/video/fb.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/video/fb.c]: @yf13 I mean passing (panelinfo.fbmem + map->offset) to vm_map_region.

Comment (xiaoxiang781216) [reply drivers/video/fb.c]: @yf13 this comment isn't addressed yet.

File: drivers/virtio/virtio-gpu.c (hunk 2 above)
Comment (xiaoxiang781216) [reply drivers/virtio/virtio-gpu.c:538]: and here too

File: drivers/video/fb.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/video/fb.c]: Yes, just mask out offset before creating virtual-physical page setup, and add offset to the returned virtual address before returrn.

Comment (yf13) [reply drivers/video/fb.c]: Can you confirm which one below is what you mean:

a) new interface is ` vm_map_region(paddr, size, offset)`? here offset can be bigger than MM_PGSIZE. 
b) new interface is `vm_map_region(paddr+offset, size)`? here offset is smaller than MM_PGSIZE.
c) none-of above, please teach explicitly.

The offset in `man mmap()` is multiple of MM_PGSIZE or zero, is this the same in NuttX?

File: drivers/virtio/virtio-gpu.c (hunk 2 above)
Comment (yf13) [reply drivers/virtio/virtio-gpu.c:538]: If fbmem is not paged aligned, then the kernel memory region before fbmem will be exposed to user space, which looks like a security hole.

File: drivers/video/fb.c (hunk 1 above)
Comment (xiaoxiang781216) [reply drivers/video/fb.c:1038]: nochangehere?

Comment (yf13) [reply drivers/video/fb.c]: rolled back to confirm  CI status...  just pushed an update supporting unaligned paddr mapping and moved `vm_map_xxx()` into ` vm_region.c`.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**